internal/pkg/cli/app_delete.go

```diff
@@ -36,51 +37,40 @@ type deleteAppOpts struct {
 	AppName          string
 
 	// Interfaces to dependencies.
-	projectService   projectService
-	workspaceService archer.Workspace
-	sessProvider     sessionProvider
-	spinner          progress
-	prompter         prompter
+	projectService       projectService
+	initProjectService   func(*deleteAppOpts) error // Overriden in test
+	workspaceService     archer.Workspace
+	initWorkspaceService func(*deleteAppOpts) error // Overriden in test
+	sessProvider         sessionProvider
+	spinner              progress
 
 	// Internal state.
 	projectEnvironments []*archer.Environment
 }
 
-func (opts *deleteAppOpts) init() error {
-	projectService, err := store.New()
-	if err != nil {
-		return fmt.Errorf("create project service: %w", err)
-	}
-	opts.projectService = projectService
-
-	workspaceService, err := workspace.New()
-	if err != nil {
-		return fmt.Errorf("intialize workspace service: %w", err)
-	}
-	opts.workspaceService = workspaceService
-
-	return nil
-}
```

Comment:
I prefer to have this `init` function because in the `project delete` workflow for example it's much simpler to call this function to initialize the `deleteAppOpts` structure rather than implementing the init funcs which don't change between instantiations.

Another alternative might be to have the `newDeleteAppOpts()` function that houses the default initialization.

Reply:
I'd prefer not to have a private function like `init` that's called from a different command. I'd expect most other commands to only have to interact with `Validate()`, `Ask()`, and `Execute()` and get initialized either with a constructor or inlined. Otherwise, a command like `project delete` would need to create a new interface with `init` in it.

So the constructor seems good to me 👍. 

Alternatively, I think in this situation I'd have initialized the `projectService` and `workspaceService` in `PreRunE` like we do in other commands. We've been initializing dependencies in `PreRunE` it can be done before `Ask()`.

Reply:
I don't think there's much difference between calling an unexported `init()` func versus setting the unexported dependency initializer funcs on the `opts` structure. There's hidden work the client has to do for the exported function calls to work at all. Wrapping those into one call (whether it be an `init()` func or constructor func) seems simpler than dealing with each one individually and could be called in the `PreRunE`.

Reply:
yeah agreed, setting unexported dependencies from the client isn't that much different. It's still slightly nicer than having an `init` method because it causes the client's interface to grow whereas it could have been handled with a constructor.

I think you settled with `init` because with a constructor that can return an error we can't initialize `opts` while building the command.

I think we should seriously consider making all our opts constructors to not return an error and call something like `log.Fatal` to exit the program if there is an error.

Reply:
Let's leave as what is now to keep consistence with other places. If we decide to refactor that in the future then we'd better change all of them together.

Reply:
:+1: for keeping it consistent to make refactoring easier across commands, but I don't think this way is consistent 😅 

Initializing these values in `PreRun` would be the consistent experience.